### DISCUSSION TITLE
File Previews

### DISCUSSION THREAD

**Discussion by kuuroo3** [directus/directus#12145] [Feature Requests]
Hey guys,

how about adding a preview for files?
Currently images can be viewed inside the Directus app but other files like PDF cannot. 
For PDF, DOC and so on you have to download the file to view it. Maybe it would be a nice addition to be able to view these files inside Directus.

Have a nice day

**Comment by benhaynes** [comment]
Absolutely, we were just talking about this last week. We could either look into a way to generate representational image previews with a library... or maybe just display them in an iframe. 🤔

**Reply by kuuroo3** [reply]
For me an iframe sounds good (enough).
This way you can display the full content and not just some snippet of a x-paged file.

**Reply by rijkvanzanten** [reply]
PDF certainly (we might even be able to rely on the browser's `<object>` element with an iframe, though we'll have to check any security implications of that).

Previewing `.docx` or `.pages` or any other proprietary files might be a little more difficult, as we would have to figure out some way to read/render the file format. For office files (word/excel/ppt etc), we might be able to rely on a third party linked service like https://docs.google.com/gview, but we'll have to be careful with potential GDPR problems with that

**Reply by benhaynes** [reply]
That link didn't work for me... but the rest I agree with!

**Reply by rijkvanzanten** [reply]
> That link didn't work for me... but the rest I agree with!

It's because it needs some more query params to point to the word file you want to preview 😁

**Reply by kuuroo3** [reply]
Agreed. PDF should be quite simple.

Aside from gdocs I also don't really have a better idea. Maybe start slow with few supported files and gear up later :)

**Reply by direnc** [reply]
Creating a PDF with the data might be a good add-on.

**Reply by codihaus** [reply]
`<iframe src='https://view.officeapps.live.com/op/embed.aspx?src=http%3A%2F%2Fieee802%2Eorg%3A80%2Fsecmail%2FdocIZSEwEqHFr%2Edoc' width='100%' height='100%' frameborder='0'>This is an embedded <a target='_blank' href='http://office.com'>Microsoft Office</a> document, powered by <a target='_blank' href='http://office.com/webapps'>Office Online</a>.</iframe>`

I think we should use Microsoft, it should be better, but yeah, preview files using GG or M$ both is ok!

**Reply by benhaynes** [reply]
If anyone wants to give a PR a try, that's would be great! I'm happy to provide a mockup design for anything we have control over. My main priority is to ensure there is nothing opinionated or proprietary with the code or design... and I don't know about relying on any external services that could go down, change, or be discontinued (which would break things in Directus).

**Reply by rijkvanzanten** [reply]
We should be really careful with relying on third party services for things like this, especially when there's no way to opt-in/out, especially especially when these services get in a lot trouble already as-is regarding GDPR (looking at Google)

**Comment by loongmxbt** [comment]
We made a [filepreview interface](https://www.npmjs.com/package/directus-extension-filepreview-interface) for temporary use, which needs to set the field to be previewed and the `token`. It can preview mp4, pdf, office files (docx, pptx, xlsx) for now.
Notice1: For office files with 3rd-party services, it can't be displayed in local environment because 3rd-party service can't read localhost files.
Notice2: .env needs to add the following:
```
CONTENT_SECURITY_POLICY_directives__child-src="'self' blob: https://view.officeapps.live.com"
```

**Reply by seymoe** [reply]
Github repo is here, if u have any doubts, submit issues please！ https://github.com/seymoe/directus-extension-filepreview-interface

**Comment by sxsws** [comment]
What are the security implications of viewing a file that could contain malicious code being rendered on the server? 

How are files that end in an extension type determined to be a genuine file of that file type?

Also, please don't use any google* services.

**Reply by rijkvanzanten** [reply]
> What are the security implications of viewing a file that could contain malicious code being rendered on the server?

Really depends on how we're viewing them. If rendered within an `iframe` through an external endpoint, we should be able to sandbox any malicious JS from running. If any inline rendering is used through vue components or otherwise, we might run into some stored xss problems.

> How are files that end in an extension type determined to be a genuine file of that file type?

The mimetype is extracted during upload, so the file-extension itself isn't used to determine the type. That being said, there's always a chance an upload is made with the wrong mimetype, especially when uploads are public. 

> Also, please don't use any google* services.

Very much agreed to that. Using any third party services from the app is tricky, let alone when it's from a privacy tricky company like Google

**Comment by rijkvanzanten** [comment]
Heya! Thanks for opening this feature request!

This feature request has received over **15 votes** from the community. This means we'll move this feature request to the Under Review state! 

The Core team will schedule a meeting to review this request as soon as possible. The discussion will then be approved or denied. You may or may not be invited to join this meeting with the core team.

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

**Comment by jaybharadia** [comment]
Any update regarding this feature ?

**Reply by rijkvanzanten** [reply]
[Unfortunately, we can not provide estimates on features or fixes](https://docs.directus.io/getting-started/support.html#can-you-give-an-eta-for-this-feature-fix). As of now, the only way to get an accurate timeline or release date is to [sponsor](https://docs.directus.io/getting-started/support.html#sponsored-work) this task. We recognize that this may be an important feature/fix, but we are a small open-source organization with a lot to triage and complete. If you can't sponsor this ticket, then the next best thing is to increase its priority by giving it a 👍

**Comment by phazonoverload** [comment]
We have recently released a few extensions via the Directus Marketplace to add more rich media previews into the Editor. We have one for PDF, one for video, and one for audio. 

With this in mind, I'll close the feature request as complete 😄
